chapters/VK_EXT_acquire_wl_display/acquire_wl_display.txt

```diff
@@ -0,0 +1,66 @@
+[open,refpage='vkAcquireWaylandDisplayEXT',desc='Acquire access to a VkDisplayKHR using Wayland',type='protos']
+--
+
+To acquire permission to directly access a display in Vulkan from a Wayland
+compositor, call:
+
+include::{generated}/api/protos/vkAcquireWaylandDisplayEXT.txt[]
+
+  * pname:physicalDevice The physical device the display is on.
+  * pname:display A wl_display connected to the Wayland compositor that
+	currently owns pname:display.
+  * pname:manager A zwp_drm_lease_manager_v1 connected to the Wayland compositor
```

Comment:
The corresponding Xlib Vulkan extension was purposely written in a manner that avoided tying it to DRM or other vendor-specific drivers.  That's why we didn't just include a lease FD in the WSI extension there as Sebastian is suggesting here (Though that would have worked fine for NV and DRM drivers), and it makes me wonder if including DRM-specific wayland protocol structs in this extension is the right direction.  For Wayland, both NV and Mesa stacks use DRM for display enumeration purposes right now, but it's been pointed out to me, and I've pointed out before that one could theoretically write a Wayland implementation entirely on top of Vulkan D2D, OpenWF, etc.  Hence, I'd prefer not to bake current de-facto standards/assumptions into the Vulkan spec if possible.

As a straw-man for discussion, would it make sense to instantiate the leasing manager object inside the Vulkan WSI implementation instead and reference the displays directly via their Vulkan handles at the API level, or is that problematic?  FWIW, the additional "translate RandR to VkDisplayKHR" entry point fell out of a similar design decision on the X11 side.

Reply:
Picking up the straw-man, I see how the WSI could try to open a Wayland connection and enumerate the leasable connectors internally, and present them via `vkGetPhysicalDeviceDisplayProperties2KHR`, without any Wayland specific extension. Whether or not that would be in the spirit of the Vulkan API, I don't know.

One issue with the display enumeration API is that `VkDisplayProperties2KHR` only provides a poorly specified `displayName` (_"Generally, this will be the name provided by the display's EDID"_) to identify the display. It would be very useful to be able to match on specific EDID properties such as vendor and product id, or even serial number, to find a specific VR headset.

Another issue is that it is unclear at which point the WSI would actually request the lease. The KHR_display extension itself doesn't provide any kind of Acquire/Release functionality.

And I think the Vulkan model of enumerating planes upfront doesn't fit well with DRM leasing, as the lessee currently has no say in the matter and only is made aware of the available planes after the lease has been issued.

Reply:
The Wayland protocol provides an out of band way to get things like the EDID for doing display selection. Splitting the "RandR -> VkDisplayKHR" step into a separate one has actually [already caused us headaches implementing this for Xwayland](https://gitlab.freedesktop.org/xorg/xserver/issues/864). The design of all of these subsystems - DRM, Wayland, etc - does not lend itself well to a design which does this in two steps, at least for Wayland in particular.

Reply:
Thanks @cubanismo for providing the context and rationale.

So one problem I have with this argument is that there is no other "leasing" mechanism on linux that I'm aware of but there is multiple windowing systems and we're already at the second extension here. I can easily see the proposed wayland extension not getting adopted (mainly due to missing authentication support, I can expand on that if wanted) but e.g. a dbus interface. Is there any realistic reason why someone would support a "leasing" mechanism that's not based on DRM? Would that ever hit mainline? If it doesn't I really don't care. Either way, if someone invents another "leasing" mechanism they could also invent a new vk extension.

To recap: there already are different ways to get a lease fd but there only is one leasing mechanism.

The staw-man would not allow for e.g. receiving the fd via dbus, it only abstracts away the specific wayland protocol so I don't see much value in that. All solutions seem to have a draw-back.

> I'd prefer not to bake current de-facto standards/assumptions into the Vulkan spec if possible

I don't think that's a fair argument. If we explicitly say in the extension that the fd is a lease fd acquired from kms/drm then it's not an assumption anymore.

In the same spirit I could argue that the xlib acquire extension bakes de-facto standards into the Vulkan spec or that this wayland proposal bakes in a not-even-de-facto standard. drm/kms leasing is at least a de-facto standard.

I hope this makes my concerns understandable.

Reply:
Whether or not this protocol will be adopted is out of scope for here, we can discuss the merits of the protocol on the mailing list. It has the explicit support of several stakeholders. This patch is not suitable for merge before it has been standardized anyway, as has already been outlined in the original post.

There is no possibility whatsoever of a dbus protocol being preferred here. There will be a hard nack and a bloody fight over that, but thankfully the only person asking for it is someone with no horse in the race.

Reply:
> There is no possibility whatsoever of a dbus protocol being preferred here.

For you maybe. Just like I won't ever prefer a wayland protocol which gives out access to exclusive resources as long as there is no compelling solution for authentication and flatpak/xdg-portals support.

All of that just supports my stance that specifying the IPC mechanism/protocol in the leasing extension seems to be a worse idea than specifying the leasing mechanism (drm/kms).

Reply:
Your preference has no bearing on the matter. You're incorrect and I don't care to further argue with you about the security of Wayland - this is *not* the place.

I would really appreciate it if you stopped derailing this discussion.

Reply:
You're just trolling me at this point. There is a CoC, behave appropriately.

Reply:
> Picking up the straw-man, I see how the WSI could try to open a Wayland connection and enumerate the leasable connectors internally, and present them via vkGetPhysicalDeviceDisplayProperties2KHR, without any Wayland specific extension.

See below.  All the window-system integration (WSI) extensions were designed to avoid drivers creating internal secret connections to the windowing systems, or even using app-provided connections outside of the scope of the APIs that take those native objects as parameters, directly or indirectly via a VkSurface reference.  Everything that isn't WSI isn't supposed to even know a window system exists.  If it doesn't reference a native window system directly, via a VkSurfaceKHR object directly, or via a swapchain derived from a VkSurfaceKHR object, it's intended to be window-system agnostic, and be implemented on top of client-side driver functionality (E.g., ioctls on device nodes, sysfs inspection, etc.) only.

> One issue with the display enumeration API is that VkDisplayProperties2KHR only provides a poorly specified displayName ("Generally, this will be the name provided by the display's EDID") to identify the display. It would be very useful to be able to match on specific EDID properties such as vendor and product id, or even serial number, to find a specific VR headset.

The core display enumeration API is indeed extremely minimalist on purpose.  It is, however, trivially extensible.  If you'd like to add an extension that returns other properties of displays, I think that would be a great idea.

> Another issue is that it is unclear at which point the WSI would actually request the lease. The KHR_display extension itself doesn't provide any kind of Acquire/Release functionality.

Right.  Displays aren't intended to be enumerated from a window system.  They're supposed to be enumerated from the device driver directly.  There has been much debate over whether this meshes well with DRM's by-design inability to enumerate displays associated with a render node, whether an application should even care about the physical displays associated with a render node/VkPhysicalDevice Vs. the displays available to it via a window system, etc.

The current design of KHR_display and the derivative leasing mechanisms is that the display leases are needed only to set a mode and present to a display, not to enumerate its properties.  However, I'd be happy to discuss any proposals that make this play better with DRM semantics.  I don't have a proposal of my own here at the moment.

> And I think the Vulkan model of enumerating planes upfront doesn't fit well with DRM leasing, as the lessee currently has no say in the matter and only is made aware of the available planes after the lease has been issued.

That's an interesting note, and ties into the above response somewhat.  Always happy to hear suggestions to improve the Vulkan display APIs.

> The Wayland protocol provides an out of band way to get things like the EDID for doing display selection. Splitting the "RandR -> VkDisplayKHR" step into a separate one has actually already caused us headaches implementing this for Xwayland. The design of all of these subsystems - DRM, Wayland, etc - does not lend itself well to a design which does this in two steps, at least for Wayland in particular.

I don't fully follow the complications in the linked issue.  Is the issue that Xwayland provides inaccurate modes for the connector and hence the application can't reliably build up its mode list prior to acquiring the lease?  Why isn't this considered a bug in Xwayland?  I recall there was considerable debate over Xrandr behavior in Xwayland environments, but I admit I didn't read most of it.  Perhaps I'm kicking dead horses.

However, I don't think the VkDisplayKHR-on-Xrandr-on-Xwayland path nor the new lease extension you're designing here need to be confined to Xwayland's RandR behavior.  You're defining custom wayland protocol already.  Presumably that protocol could be used to tunnel an accurate mode list, EDID, etc. back to the Vulkan implementation.  Xwayland could use this path rather than its usual RandR implementation when sending modes to the Vulkan driver as opposed to general RandR client applications.

BTW, I don't know how valid the concerns in the other issue about current behavior being baked into games are.  From my understanding, this behavior is only baked into VR intermediary libraries like the SteamVR runtime, which are likely much more easily updated.  However, this is speculation.  We'd have to go check the code/ask the platform owners.  I for one wouldn't be opposed to a larger re-design if there's a more suitable API for leasing, assuming the users would be on board.

> To recap: there already are different ways to get a lease fd but there only is one leasing mechanism.

There are two.  NVIDIA has a different leasing mechanism than DRM.  It still uses FDs to represent the lease.  For Wayland, it's not clear yet, but we'd likely use the DRM leasing mechanism even in the NVIDIA proprietary driver, so we're interested in the semantics of both.  How we behave in a hypothetical future where our X DDX is running in an Xwayland instance without direct access to our display hardware, but where we still want to maintain compatibility with applications that rely on our prior X DDX behavior is also an interesting line of thought.

In general, I'm rarely swayed by any reasoning of the form "If it doesn't exist now, on Linux, let's pretend it never will, anywhere," especially when it comes to designing extensible APIs that have a habit of being baked into applications that are difficult to update.  I noted above that we may still have a window of time to modify the existing Vulkan leasing APIs without major repercussion if there's sufficient need, but that window could close at any time, so I'd rather not rely on it as a design principle.

Reply:
> In general, I'm rarely swayed by any reasoning of the form "If it doesn't exist now, on Linux, let's pretend it never will, anywhere,"

The same argument is true for yet another way to receive a lease fd. The question here boils down to this: do we want to abstract away to way your receive the fd or do we want to abstract away what you receive.

Both could change and both already seem to have two answers (x11 vs wayland and drm fd vs nvidia fd).

Reply:
> See below. All the window-system integration (WSI) extensions were designed to avoid drivers creating internal secret connections to the windowing systems, or even using app-provided connections outside of the scope of the APIs that take those native objects as parameters, directly or indirectly via a VkSurface reference. Everything that isn't WSI isn't supposed to even know a window system exists.

Thank you for explaining the intention in detail, I never got that just from reading the spec. Assuming that indeed the only way to get a `wl_display` into Vulkan as intended would be via `vkCreateWaylandSurfaceKHR`, it seems weird then that an application that doesn't want to render into a Wayland surface at all would have to create an otherwise unused Wayland surface, just to be able to request a lease from the compositor.

> If it doesn't reference a native window system directly, via a VkSurfaceKHR object directly, or via a swapchain derived from a VkSurfaceKHR object, it's intended to be window-system agnostic, and be implemented on top of client-side driver functionality (E.g., ioctls on device nodes, sysfs inspection, etc.) only.

So the idea is that `vkGetPhysicalDeviceDisplayPropertiesKHR` lists all displays, even those that can't be used (because the compositor won't lease them to us)? Similarly, since it is the Wayland compositor's decision which planes to lease to us, `vkGetDisplayPlaneSupportedDisplaysKHR` can't possibly know which hardware planes are available before the lease is requested.

> > One issue with the display enumeration API is that VkDisplayProperties2KHR only provides a poorly specified displayName ("Generally, this will be the name provided by the display's EDID") to identify the display. It would be very useful to be able to match on specific EDID properties such as vendor and product id, or even serial number, to find a specific VR headset.
> 
> The core display enumeration API is indeed extremely minimalist on purpose. It is, however, trivially extensible. If you'd like to add an extension that returns other properties of displays, I think that would be a great idea.

Ok, I see this is not an issue / out of scope in this context.

> > Another issue is that it is unclear at which point the WSI would actually request the lease. The KHR_display extension itself doesn't provide any kind of Acquire/Release functionality.
> 
> Right. Displays aren't intended to be enumerated from a window system. They're supposed to be enumerated from the device driver directly.

Say the `VkDisplayKHR` are enumerated from a DRM modesetting node if available. Given such a display, which is not leased yet, at what point should the actual lease be requested? When creating the display surface? When creating the swapchain? When presenting to it for the first time, as that's probably what triggers the modeset?

> There has been much debate over whether this meshes well with DRM's by-design inability to enumerate displays associated with a render node, whether an application should even care about the physical displays associated with a render node/VkPhysicalDevice Vs. the displays available to it via a window system, etc.

Good point, what if the application does not have access to a modesetting node at all until it is provided with one via the leasing mechanism?

> The current design of KHR_display and the derivative leasing mechanisms is that the display leases are needed only to set a mode and present to a display, not to enumerate its properties. However, I'd be happy to discuss any proposals that make this play better with DRM semantics. I don't have a proposal of my own here at the moment.

How about an extension structure to `VkDisplaySurfaceCreateInfoKHR` that allows to pass a Wayland `VkSurfaceKHR` (which internally provides the `wl_display`) to request the lease from?

Reply:
> How about an extension structure to VkDisplaySurfaceCreateInfoKHR that allows to pass a Wayland VkSurfaceKHR (which internally provides the wl_display) to request the lease from?

I think in my reply above I left out the case where an entry point directly operates on a native display server connection above.  I think it's fine to have calls like vkDoWaylandSpecificVulkanStuffEXT(wl_display *wl_dpy, ...) that do wayland stuff, similar to the X11-specific commands.  I honestly think the abstraction of putting VkSurface into the API as a real Vulkan object type was a mistake in retrospect, but the intention there was allow more of the swapchain API to be winsys-agnostic.  Perhaps there should have been a VkLessor abstraction introduced for the X11 leasing or something, but it wasn't clear at the time how much API commonality could have been derived from that.

Regardless, I think the important thing is that Vulkan shouldn't have to spin up its own display/compositor connections behind the applications back, nor send X11 or compositor protocol outside of commands that operate on corresponding native objects directly or indirectly.  These are the sorts of things that made EGL and GLX-based GL drivers complicated and prone to nasty things like surprise XGrabServer() deadlocks.  It could be there are already mistakes that result in cases where this is required, but if there are, they're indeed mistakes IMHO, not by design.

Reply:
I know I'm why late ont this, but for random reasons stumbled over this and figured I'll throw the upstream DRM perspective in. For a normal client (i.e. not a compositor or something running on bare metal which either has special access rights, or knows how and has permission to get display access from systemd-logind) it's impossible to enumerate displays on the DRM stack. Two reasons:
- properly locked down clients have only access to /dev/dri/renderD*, which does not give access to any display resources. We're also not going to add new kernel<->userspace abi that assumes locked down client isn't a use-case, so direct enumeration of display resources wont happen in the future.
- gl/egl/glx can get a /dev/dri/card* device (at least sometimes, but e.g. the amd mesa stack has switched over to render nodes in all cases), but again generally the only way to get it is through winsys protocol. So definitely not the vulkan way, and all mesa vulkan drivers just enumerate /dev/dri/renderD*.
- on multi-gpu often the display device isn't on the same card* device as the rendering is done, so even if you happen to have access to a card* good chances it's the wrong one. This is the case for most optimus style laptop setups, and all arm SoC platforms except qualcomm and some very old tegra.

Only way to enumerate direct display devices is through winsys protocol. Afaik the current vk direct display implemenation for X11 in mesa just makes the behind-the-scenes winsys connection. Totally breaks the spirit of vk, but gets the job done. I think the same is required on wayland and everywhere else really.

In practice given that it's all done & shipping I don't think there's much we can do now. So *shrug*.

Some historical notes maybe: I think this didn't come up because first versions of the mesa x11 direct display support had hacks to just enumerate directly using /dev/dri/card*. But that died in review before it all landed.